jax/lax/lax.py

```diff
@@ -3612,68 +3612,100 @@ def _select_and_gather_add_shape_rule(
-
-  def fst(t):
-    st = c.ConvertElementType(c.ShiftRightLogical(t, bits_const), uint_etype)
-    return c.BitcastConvertType(st, etype)
-  sx = fst(x)
-  sy = fst(y)
-
-  assert select_prim is ge_p or select_prim is le_p
-  which = c.Ge(sx, sy) if select_prim is ge_p else c.Le(sx, sy)
-  c.Select(which, x, y)
-  return c.Build()
 
 def _select_and_gather_add_translation(
     c, tangents, operand, select_prim, window_dimensions, window_strides,
-    padding):
+    padding, max_bits=64):
   # XLA doesn't yet implement ReduceWindow on tuples (Google bug b/73062247), so
   # we implement a pair-wise ReduceWindow by packing two k-bit values into
   # 2k-bit unsigned integer using bit tricks. This will only work for <= 32-bit
   # inputs (since we don't have 128-bit integer types).
-  dtype = c.GetShape(operand).numpy_dtype()
-  bits = onp.finfo(dtype).bits
-  if bits > 32:
-    raise NotImplementedError(
-        "select_and_gather_add is not implemented for type larger than 32 bits")
-  etype = xla_bridge.dtype_to_etype(dtype)
-  uint_etype = xla_bridge.dtype_to_etype(_UINT_DTYPES[bits])
-  pair_uint_dtype = _UINT_DTYPES[bits * 2]
-  pair_uint_etype = xla_bridge.dtype_to_etype_exact(pair_uint_dtype)
-
-  operand = c.BitcastConvertType(operand, uint_etype)
-  tangents = c.BitcastConvertType(tangents, uint_etype)
-  operand = c.ConvertElementType(operand, pair_uint_etype)
-  tangents = c.ConvertElementType(tangents, pair_uint_etype)
-  operand = c.ShiftLeft(
-    operand, c.Constant(pair_uint_dtype(bits), canonicalize_types=False))
+  shape = c.GetShape(operand)
+  dtype = shape.numpy_dtype()
+  etype = shape.xla_element_type()
+  nbits = onp.finfo(dtype).bits
+
+  assert nbits <= max_bits
+  double_word_reduction = nbits * 2 <= max_bits
+
+  const = lambda c, dtype, x: c.Constant(onp.array(x, dtype=dtype),
+                                         canonicalize_types=False)
+
+  if double_word_reduction:
+    word_dtype = _UINT_DTYPES[nbits]
+    double_word_dtype = _UINT_DTYPES[nbits * 2]
+    word_type = xla_bridge.dtype_to_etype_exact(word_dtype)
+    double_word_type = xla_bridge.dtype_to_etype_exact(double_word_dtype)
+
+    # Packs two values into a tuple.
+    def pack(a, b):
+      a = c.BitcastConvertType(a, word_type)
+      b = c.BitcastConvertType(b, word_type)
+      a = c.ConvertElementType(a, double_word_type)
+      b = c.ConvertElementType(b, double_word_type)
+      a = c.ShiftLeft(a, const(c, double_word_dtype, nbits))
+      return c.Or(a, b)
+
+    # Unpacks the first element of a tuple.
+    def fst(c, t):
+      st = c.ShiftRightLogical(t, const(c, double_word_dtype, nbits))
+      return c.BitcastConvertType(c.ConvertElementType(st, word_type), etype)
+
+    # Unpacks the second element of a tuple.
+    def snd(t):
+      return c.BitcastConvertType(c.ConvertElementType(t, word_type), etype)
+
+  else:
+    warnings.warn("Using reduced precision for gradient of reduce-window "
```

Comment:
Maybe add a reference to b/73062247 here too?

Reply:
Done.